android/core/src/main/java/ai/djl/android/core/BitmapImageFactory.java

```diff
@@ -86,24 +86,29 @@ public Image fromImage(Object image) {
     @Override
     public Image fromNDArray(NDArray array) {
         Shape shape = array.getShape();
+        int height = (int) shape.get(1);
+        int width = (int) shape.get(2);
         if (shape.dimension() != 3) {
             throw new IllegalArgumentException("Shape should only have three dimension follow CHW");
         }
         if (shape.get(0) == 1) {
             throw new UnsupportedOperationException("Grayscale image is not supported");
+        } else if (shape.get(0) != 3) {
+            if (shape.get(2) == 3) {
+                height = (int) shape.get(0);
+                width = (int) shape.get(1);
+            } else {
+                throw new IllegalArgumentException("First or last dimension should be number of channel with value 1 or 3");
+            }
         }
-        if (NDImageUtils.isCHW(shape)) {
-            array = array.transpose(2, 0, 1);
-            shape = array.getShape();
-        }
-        int height = (int) shape.get(1);
-        int width = (int) shape.get(2);
-        int imageArea = width * height;
-        Bitmap bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
-        int[] raw = array.toType(DataType.UINT8, false).toUint8Array();
+        int finalWidth = width;
+        int finalHeight = height;
+        int imageArea = finalWidth * finalHeight;
+        Bitmap bitmap = Bitmap.createBitmap(finalWidth, finalHeight, Bitmap.Config.ARGB_8888);
+        int[] raw = array.toUint8Array();
         IntStream.range(0, imageArea).parallel().forEach(ele -> {
-            int x = ele % width;
-            int y = ele / width;
+            int x = ele % finalWidth;
```

Comment:
This does not applies to HWC. HWC (Channel last) will display the data like 

For each height and width:
Show R,G,B (3 channel color)

CHW like

For each channel
 Hight and width color

Reply:
So you may need to determine if here is CHW or HWC and customize your loading logic

Reply:
I have tested this on all the Android apps and this method works. Originally for my app, I was transposing the NDArray to CHW, but that caused an error in the image being shown in the app. @frankfliu and I were getting a 3x3 grid of mini grayscale images back, but when I took out the transpose, the image was shown correctly.

Reply:
Since you are allowing HWC data input, you should handle this case correctly. The data order is changed for HWC as mentioned above. It is ordered to be channel last where RGB color are come together. In here we assume data is CHW and that's why we always add the area size to skip a channel